Chat Private

jazzz opened the pull request:
This PR adds a specification which describes a conversation protocol for pairwise communication called PRIVATE1. 

It covers payload generation and operation after a key has been agreed upon by both parties.

### Notes 
This may eventually be simplified and use [Reliable Channels](https://github.com/waku-org/specs/pull/89) but its explicit in how it utilizes SDS + segmentation and the reasons for doing so.

jazzz commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+
+The following mappings connect PRIVATE1 concepts to SDS fields:
+
+- `sender_id`: !TODO: This requires PRIVATE1 to be identity aware
```
Unsure exactly what the best course of action is here. Recent changes to SDS requires an sender_id be attached the the SdsMessage. 

This then brings in a concept of identity to this specification. Its definitely possible to introduce Identity concepts in order to pass them through to SDS, however it seems like a lot of added complexity

kaichaosun commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+PRIVATE1 uses bloom filter parameters of `n=2000` (expected elements) and `p=0.001` (false positive probability).
+This configuration produces bloom filters of approximately 3.5 KiB per message.
+
+!TODO: Can the bloom filter be dropped in 1:1 communication?
```
I agree with the split.
Only use casual history is enough for message sent check.

jonesmarvin8 commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+- **[DOUBLERATCHET]** "The Double Ratchet Algorithm", Signal, 2016.
+  https://signal.org/docs/specifications/doubleratchet/
+
+- **[SDS]** "Scalable Data Sync Specification", vac, 2024.
```
All of these github references from vacp2p and waku-org are broken.

osmaczko commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+Implementation specifics:
+- Error correction is not used, as reliable delivery is already provided by lower layers. 
+- `segmentSize` = `max_seg_size`
+- All payloads regardless of size are wrapped in a segmentation message.
```
This contradicts segmentation specs:
> Messages whose payload size is ≤ segmentSize are sent unmodified.

osmaczko reviewed (approved):
Looks good overall. The reasoning behind the `Content -> Segmentation -> Reliability -> Encryption` layering makes sense.

Please remember to address the TODOs before merging.

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+
+This specification defines PRIVATE1, a conversation protocol for establishing secure, full-duplex encrypted communication channels between two participants. PRIVATE1 provides end-to-end encryption with forward secrecy and post-compromise security using the Double Ratchet algorithm, combined with reliable message delivery via Scalable Data Sync (SDS) and efficient segmentation for transport-constr ...
+
+The protocol is transport-agnostic and designed to support both direct messaging and as a foundation for group communication systems. PRIVATE1 ensures payload confidentiality, content integrity, sender privacy, and message reliability while remaining resilient to network disruptions and message reordering.
```
I'm not sure but something sounds wrong in `... support both direct messaging and as a foundation ...`. I get the meaning, just not sure of the language.

I think you know better, I'm just highlighting
```suggestion
The protocol is transport-agnostic and designed to support direct messaging and serve as a foundation for group communication systems. PRIVATE1 ensures payload confidentiality, content integrity, sender privacy, and message reliability while remaining resilient to network disruptions and message reordering.
```

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+
+## Definitions
+
+This document makes use of the shared terminology defined in the [CHAT-DEFINITIONS](https://github.com/waku-org/specs/blob/jazzz/chatdefs/informational/chatdefs.md) specification.
```
```suggestion
This document makes use of the shared terminology defined in the [CHAT-DEFINITIONS](informational/chatdefs.md) specification.
```

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+
+### Payload Delivery
+How payloads are sent and received by clients is deliberately not specified by this protocol.
+Transport choice is an implementation decision that should be made based on deployment requirements.
```
Am I reading this right? Or does it actually mean "PrivateV1 implementation"?
```suggestion
Transport choice is an application decision that should be made based on deployment requirements.
```

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+- `rsk` and `ssk` SHOULD incorporate ephemeral key material
+
+PRIVATE1 requires a unique identifier, however the exact derivation is left to implementations to determine.
+- `conversation_id` MUST be unique across all instances of chat conversations
```
What is meant by `chat` here? And what is `chat conversations`?

igor-sirotin commented on standards/application/privatev1.md:
````diff
@@ -0,0 +1,417 @@
+This conversation type assumes there is some service or application which wishes to generate and receive end-to-end encrypted content. 
+It also assumes that some other component is responsible for delivering the generated payloads. At its core this protocol takes the content provided and creates a series of payloads to be sent to the recipient.
+
+```mermaid
+flowchart LR
+    Content:::plain--> PrivateV1 --> Payload:::plain
+    classDef plain fill:none,stroke:transparent;
+```
````
I agree that Frames seem to be unnecessary here.

Maybe something like this?
```mermaid
flowchart LR
    Content:::plain--> PrivateV1 --> Payloads@{ shape: stacked-rectangle }
    classDef plain fill:none;
    class Payloads plain;
```

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+- `conversation_id` SHOULD be consistent across applications to maintain interoperability
+
+Additionally implementations MUST determine the following constants:
+- `max_seg_size` - maximum segmentation size to be used.
```
Isn't this dictated by the transport used by an application? 
If yes, then this `max_seg_size` should be defined by application, not protocol implementation? Or am I not getting "implementation" term here?

jazzz replied:
>  Isn't this dictated by the transport used by an application?

When this was modeled it implemented relaibility inside of the converastion. 

> then this max_seg_size should be defined by application, n

sure. However we're trying to avoid a circular dependency. `max_seg_size` is a paramter that must be passed in.

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+- `conversation_id` SHOULD be consistent across applications to maintain interoperability
+
+Additionally implementations MUST determine the following constants:
+- `max_seg_size` - maximum segmentation size to be used.
```
If `segmentSize` is exactly `max_seg_size`, should we mention that actual payloads can and will be over larger than this limit?

jazzz replied:
No. Thats covered at a higher level. the `max_seg_size` doesn't have to match the MTU of the transport layer. Its just a parameter. 

The WakuChat protocol spec would force these constraints.

igor-sirotin commented on standards/application/privatev1.md:
```diff
@@ -0,0 +1,417 @@
+**Processing Rules**
+
+- All application-level content MUST use the `content` frame type
+- Clients SHALL only pass `content` frames to applications
+- Clients MAY drop unrecognized frame types
```
Is "clients" same thing as "implementations"?
```suggestion
- Implementations SHALL only pass `content` frames to applications
- Implementations MAY drop unrecognized frame types
```

igor-sirotin reviewed (approved):
I wish I read it earlier!

jazzz committed:
PrivateV1

Cleanup of Arch, Initialization and Segmentation

frame handling + initialization

towards semBr

Abstract away contentFrame

update  frame handing

Remove delivery address

remove partial order property

Update background

readability updates

Update encryption and segmentation

Remove stale section

update payload delivery warnings

Update SDS initialization

Update security considerations

Cleanup

review comments

spelling fixes

fix: pyspelling config

Update DR params

Update standards/application/privatev1.md

Co-authored-by: Igor Sirotin <igor.sirotin.1012@gmail.com>

Clarify implementation

jazzz committed:
Merge branch 'master' of github.com:waku-org/specs into jazzz/private
